src/Microsoft.Health.Fhir.Core/Features/Persistence/BadRequestException.cs

```diff
@@ -27,7 +29,12 @@ public BadRequestException(string errorMessage)
         /// <param name="errorMessages">Collection of error strings that are used to initialize OperationOutcomeIssue collection.
         /// Must be non-null and non-empty</param>
         public BadRequestException(ICollection<string> errorMessages)
-            : base((errorMessages == null || errorMessages.Count < 2) ? errorMessages.First() : "Multiple bad request errors.")
+            : base(((Func<string>)(() =>
+            {
+                Debug.Assert(errorMessages != null, "Parameter errorMessages must not be null.");
+                Debug.Assert(errorMessages == null || errorMessages.Any(), "Parameter errorMessages must not be empty.");
+                return (errorMessages == null || errorMessages.Count < 2) ? errorMessages.First() : "Multiple bad request errors."; // "Multiple bad request errors." message is used only internally, no need to add it to the Resources.resx file.
```

Comment:
Is there a reason why we have the check for `errorMessages == null` on lines 35 and 36? It looks like we cover this case on line 34.

Reply:
It's not the same, one is != other is ==. According to documentation Debug.Assert will put out debug message and then continue with execution. If errorMessages is null, then first DebugAssert will put the message "must not be null", and then execution will continue to the second Debug.Assert. If I didn't put that second errorMessages == null, then it would try Any() on null reference and throw exception.

So it's all because Debug.Assert continues to execute after the test, depending on how it's setup. https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assert

Reply:
Ah, it is my understanding that `Debug.Assert()` will halt execution and return the call stack.

I am reading this as:
Line 34 - if `errorMessages` is null, halt execution. Otherwise proceed to line 35.

Reply:
Doc says if app has user interface it will display message box with stack and options ignore, retry, abort. Clicking ignore will continue execution.

If app has no user interface it only says it will send the message to the listeners. Remarks: "If the result is false, it sends the specified diagnostic message to the Listeners collection." That's all it says.

It's here https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assert under "Remarks" for any Assert method

Reply:
System.Diagnostics.Debug.Assert(false, "helloA");
            System.Diagnostics.Debug.Assert(false, "helloB");

Doc says A will only output message.
Console test app executes B.
FHIR-server when driven by unit tests throws at A and does not reach B
FHIR-server when started standalone and invoked through rest executes both A and B

So, second test stays.

Reply:
Thanks - I didn't know `Debug.Assert()` will only halt execution when running the tests and not when running the server standalone. I created a user story to investigate how we might be able to make this consistent [AB#86744](https://microsofthealth.visualstudio.com/Health/_workitems/edit/86744).